### PR TITLE
quick-fix for --gpus flag bug

### LINKED ISSUE #2669: --gpus flag with add_argparse_args bug
## 🐛 Bug

When using

```
parser = Trainer.add_argparse_args(parser)
args = parser.parse_args()

trainer = Trainer.from_argparse_args(args)
```

if the user does not provide the --gpus flag, the code allocates some memory on the GPU and reports

GPU available: True, used: True

but does not use this GPU. This issue has been discovered in bolts: 

1. https://github.com/PyTorchLightning/pytorch-lightning-bolts/issues/35
2. https://github.com/PyTorchLightning/pytorch-lightning-bolts/issues/124

### To Reproduce

Use the following method to create the trainer object

```
parser = Trainer.add_argparse_args(parser)
args = parser.parse_args()

trainer = Trainer.from_argparse_args(args)
```

and don't pass the --gpus flag.

### Expected behavior

If --gpus flag is not provided in the script call, then Lightning should report 

GPU available: True, used: False 

with a warning, and not allocate any memory on the GPU. This way, the user can set the --gpus flag if they have missed out on it.

### PR DESCRIPTION
**Pull request by ananyahjha93**
## What does this PR do?

Prevents setting self.on_gpu as True when --gpus flag is not passed by the user in the case where the following way is used to create the trainer
```
parser = Trainer.add_argparse_args(parser)
args = parser.parse_args()

trainer = Trainer.from_argparse_args(args)
```

Fixes #2669 

# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together? Otherwise, we ask you to create a separate PR for every change.
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? 
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?

### CODE REVIEW

**Comment by awaelchli** [comment]
I think the true fix should be to parse the gpus flag first and then set the on_gpu property, right?

**Comment by ananyahjha93** [comment]
@awaelchli In this case type takes a callable as a value in form Trainer._arg_default which then evaluates to the gpus flag values if something is passed to the script by the user. If the user does not include the --gpus flag, args.gpus refers to Trainer._arg_default callable.

https://github.com/PyTorchLightning/pytorch-lightning/blob/bugfix/gpus-flag/pytorch_lightning/trainer/trainer.py#L786

**Comment by awaelchli** [comment]
Yes I understand that. We also have a function that takes the user input and converts it to a list. 
See distrib_parts.py::_parse_gpu_ids. 
There, a callable results in None, which is what we want, right?
It happens later in the trainer init, so if you just move the on_gpu down there after it was parsed, it should be fine, now?

**Comment by Borda** [comment]
well, I would not pass the parsing function at all... it makes unnecessary complex

**Comment by Borda** [comment]
yes, I would not pass the function at all, it makes unnecessary complex...

**Comment by ananyahjha93** [comment]
@awaelchli yeah, you are right, thanks for pointing this out.

@Borda what's your suggestion here, if not pass a function?

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -532,6 +532,10 @@ def __init__(
         self.root_gpu = determine_root_gpu_device(self.data_parallel_device_ids)
         self.root_device = torch.device("cpu")
 
+        # self.data_parallel_device_ids is None if gpus is callable
```
Comment (williamFalcon) [inline pytorch_lightning/trainer/trainer.py:535]: i don't understand what this means

Comment (ananyahjha93) [reply pytorch_lightning/trainer/trainer.py:535]:
```
self.data_parallel_device_ids = _parse_gpu_ids(self.gpus)
```
_parse_gpu_ids returns None is self.gpus is callable, if you don't pass --gpus flag then self.gpus points to Trainer._arg_default

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:535]: We could move the `self.on_gpu = True if (gpus and torch.cuda.is_available()) else False` found a few lines above
down here and simply write

`self.on_gpu = True if (self.data_parallel_device_ids and torch.cuda.is_available()) else False`

then there is no need for an extra if clause or a comment :)

**Comment by Borda** [comment]
we shall re-count nb tests we are running as this was merged automatically even one test was failing...

**Pull request merged**